### PR TITLE
various memory leaks plugged

### PR DESCRIPTION
**Pull request by christopherlam**
~instead of `qof_print_date` string which must be freed.~

### CODE REVIEW

**Comment by christopherlam** [comment]
This would leak numerous datebuff strings, 1 per register date-cell.

**Comment by jralls** [comment]
@christopherlam Looks good.

**Comment by christopherlam** [comment]
https://stackoverflow.com/questions/52523831/gvalue-initialization-finalization-necessity

**Comment by christopherlam** [comment]
(see answer by philip withnall, the gnome dev IIRC)

File: libgnucash/engine/Transaction.c [hunk 1]
```diff
@@ -2472,15 +2481,18 @@ xaccTransGetTxnType (const Transaction *trans)
...
     if (!trans) return TXN_TYPE_NONE;
     qof_instance_get_kvp (QOF_INSTANCE (trans), &v, 1, TRANS_TXN_TYPE_KVP);
     if (G_VALUE_HOLDS_STRING (&v))
+    {
          s = g_value_get_string (&v);
-    if (s && strlen (s) == 1)
-        return *s;
-
-    return TXN_TYPE_NONE;
+         if (s && (s[0] != '\0') && (s[1] == '\0'))
```
Comment (gjanssens) [inline libgnucash/engine/Transaction.c:2491]: Why this change ?

Comment (christopherlam) [reply libgnucash/engine/Transaction.c:2491]: Small micro-optimisation. `strlen` is a function call and is O(N), whereas this translates to 6 assembly instructions...

Comment (gjanssens) [reply libgnucash/engine/Transaction.c:2491]: Fair enough.

On the other hand the original code expresses the intent much more clearly where your version requires mental decoding to figure out we are trying the first character of a one-byte string, only if the string is exactly one byte. So unless you have measurable performance gains from this I would favour the more expressive form.

Just a nitpick though.

Comment (gjanssens) [reply libgnucash/engine/Transaction.c]: Indeed.

We're not in C++ land yet, but would there be a C++ alternative not using macros ? I hear a lot of backpush against macros in C++ conversations.

File: libgnucash/engine/Transaction.c [hunk 2]
```diff
@@ -2472,15 +2481,18 @@ xaccTransGetTxnType (const Transaction *trans)
...
+             ret = s[0];
+    }
+    g_value_unset (&v);
```
[the remark this answers is not in the document]
Comment (gjanssens) [reply libgnucash/engine/Transaction.c:2494]: I *think*  'g_value_take_string' works the other way around. It would be used to set a GValue with the given string such that it takes ownership of it. It's not a method to extract the string (a glaring omission in the GValue specification in retrospect).

[the remark this answers is not in the document]
Comment (gjanssens) [reply libgnucash/engine/Transaction.c:2494]: https://developer.gnome.org/gobject/stable/gobject-Standard-Parameter-and-Value-Types.html#g-value-take-string

[the remark this answers is not in the document]
Comment (jralls) [reply libgnucash/engine/Transaction.c:2494]: Darn, you're right.

File: libgnucash/engine/Transaction.c (hunk 1 above)
Comment (jralls) [reply libgnucash/engine/Transaction.c]: Sort of, and it works in C too:
```
static inline char 
str_one_char (const char* s) 
{
   return (s && s[0] && !s[1]) ? *s : '\0'; 
}
```
like a macro the *definition* has to be part of the compilation unit (so it's either in the source file being compiled or in a header) or it can't be inlined. Also `inline` is a suggestion, the compiler doesn't have to inline it and probably won't if told `-O0` and if it doesn't inline it then the string has to be pretty long before it improves on `strlen`.

Comment (christopherlam) [reply libgnucash/engine/Transaction.c]: let's resolve this. `strlen` it is then.

File: libgnucash/engine/qof-backend.cpp [hunk 3]
```diff
@@ -100,8 +100,9 @@ QofBackend::register_backend(const char* directory, const char* module_name)
     }
 
     auto absdir = directory;
+    auto pkgdir = gnc_path_get_pkglibdir ();
     if (!absdir || !g_path_is_absolute(absdir))
-        absdir = gnc_path_get_pkglibdir ();
+        absdir = pkgdir;
```
Comment (jralls) [inline libgnucash/engine/qof-backend.cpp:105]: Why?
